File: src/transport.ts
/**
 * TransportApi — the only network surface (spec §6, §7).  verify() never
 * enters it.  POSTs canonical JSON to the configured registry origin only;
 * no redirects are followed and no URLs found in artifacts are resolved.
 */

import { jcsBytes } from "./encoding/jcs.js";
import { parseStrictJson } from "./encoding/strict-json.js";
import { SunlightError, httpStatusFor, type ErrorCode } from "./errors.js";
import { newId } from "./ids.js";
import { RAW } from "./crypto/domains.js";
import type { HeadRef } from "./schema.js";

export interface TransportOptions {
  origin: string;
  bearer: string;
  connectTimeoutMs: number;
  readTimeoutMs: number;
  requestId?: () => string;
}

const BODY_CAP = 32 * 1024;
const KNOWN = new Set([
  "BAD_REQUEST", "SCHEMA_INVALID", "VERSION_UNSUPPORTED", "METHOD_UNKNOWN",
  "ID_MISMATCH", "UNAUTHENTICATED", "FORBIDDEN", "KEY_INACTIVE",
  "ROLE_MISMATCH", "NOT_FOUND", "HEAD_CONFLICT", "CUT_MISMATCH",
  "IDEMPOTENCY_CONFLICT", "STATEMENT_EXISTS", "ID_CONFLICT", "STATE_CONFLICT",
  "BODY_LIMIT", "BUNDLE_LIMIT", "GRAPH_LIMIT", "CONTENT_TYPE",
  "CONTENT_ENCODING", "HASH_MISMATCH", "SIGNATURE_INVALID", "PARENT_MISSING",
  "PARENT_MISMATCH", "PARENT_INACTIVE", "DESCRIPTOR_MISMATCH",
  "STATEMENT_INVALID", "COMMAND_EXPIRED", "CLOCK_AHEAD", "RATE_LIMIT",
  "STORAGE_UNAVAILABLE", "AUDIT_KEY_UNAVAILABLE", "CAPACITY_LIMIT",
]);

export class RegistryTransport {
  constructor(private readonly opts: TransportOptions) {}

  async rpc(method: string, params: unknown, id?: string): Promise<unknown> {
    const requestId = id ?? this.opts.requestId?.() ?? newId("slq");
    const req = { v: "sunlight.rpc/1", id: requestId, method, params };
    const body = jcsBytes(req);
    if (body.length > BODY_CAP) throw new SunlightError("BODY_LIMIT");

    const url = this.opts.origin.replace(/\/$/, "") + "/v1/rpc";
    const ctl = new AbortController();
    const timer = setTimeout(() => ctl.abort(), this.opts.readTimeoutMs);
    let res: Response;
    try {
      res = await fetch(url, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${this.opts.bearer}`,
        },
        body,
        redirect: "error",
        signal: ctl.signal,
      });
    } catch (e) {
      throw new SunlightError("NETWORK_ERROR", null, e);
    } finally {
      clearTimeout(timer);
    }

    const raw = Buffer.from(await res.arrayBuffer());
    let parsed: unknown;
    try {
      parsed = parseStrictJson(raw);
    } catch {
      throw new SunlightError("NETWORK_ERROR", null, new Error("unparseable response"));
    }
    if (typeof parsed !== "object" || parsed === null) {
      throw new SunlightError("NETWORK_ERROR", null, new Error("non-object response"));
    }
    const obj = parsed as Record<string, unknown>;

    // HTTP-layer error shape: {"error":{code,retryable,head}}
    if ("error" in obj && obj.v === undefined) {
      const err = obj.error as Record<string, unknown>;
      throw new SunlightError(normalizeCode(err?.code), readHead(err));
    }
    // RpcResponse
    if (obj.ok === true) return obj.result;
    if (obj.ok === false) {
      const err = obj.error as Record<string, unknown>;
      throw new SunlightError(normalizeCode(err?.code), readHead(err));
    }
    throw new SunlightError("NETWORK_ERROR", null, new Error("malformed response"));
  }
}

function normalizeCode(c: unknown): ErrorCode {
  return typeof c === "string" && KNOWN.has(c) ? (c as ErrorCode) : "NETWORK_ERROR";
}

function readHead(err: unknown): HeadRef | null {
  const h = (err as { head?: unknown })?.head;
  if (
    typeof h === "object" && h !== null &&
    typeof (h as { seq?: unknown }).seq === "number" &&
    typeof (h as { hash?: unknown }).hash === "string"
  ) {
    return h as HeadRef;
  }
  return null;
}

File: python/sunlight/transport.py
"""TransportApi — the only network surface (spec §6, §7).  verify() never
enters it.  POSTs canonical JSON to the configured registry origin only; no
redirects are followed and no URLs found in artifacts are resolved.
"""

from __future__ import annotations

import urllib.error
import urllib.request

from .errors import ERROR_CODES, SunlightError
from .ids import new_id
from .jcs import jcs
from .strictjson import parse_strict_json

BODY_CAP = 32 * 1024


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


class RegistryTransport:
    def __init__(
        self,
        origin: str,
        bearer: str,
        connect_timeout_ms: int = 5000,
        read_timeout_ms: int = 10000,
        request_id=None,
    ):
        self.origin = origin
        self.bearer = bearer
        self.read_timeout_s = max(1, read_timeout_ms) / 1000.0
        self.request_id = request_id
        self._opener = urllib.request.build_opener(_NoRedirect)

    def rpc(self, method: str, params, req_id: str | None = None):
        request_id = (
            req_id
            or (self.request_id() if self.request_id else None)
            or new_id("slq")
        )
        req = {"v": "sunlight.rpc/1", "id": request_id, "method": method, "params": params}
        body = jcs(req)
        if len(body) > BODY_CAP:
            raise SunlightError("BODY_LIMIT")

        url = self.origin.rstrip("/") + "/v1/rpc"
        http_req = urllib.request.Request(
            url,
            data=body,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.bearer}",
            },
            method="POST",
        )
        try:
            res = self._opener.open(http_req, timeout=self.read_timeout_s)
            raw = res.read(BODY_CAP + 1)
        except urllib.error.HTTPError as e:
            try:
                raw = e.read(BODY_CAP + 1)
            except Exception:
                raise SunlightError("NETWORK_ERROR", cause=e) from e
        except Exception as e:
            raise SunlightError("NETWORK_ERROR", cause=e) from e

        try:
            parsed = parse_strict_json(raw)
        except Exception as e:
            raise SunlightError("NETWORK_ERROR", cause=ValueError("unparseable response")) from e
        if not isinstance(parsed, dict):
            raise SunlightError("NETWORK_ERROR", cause=ValueError("non-object response"))

        if "error" in parsed and parsed.get("v") is None:
            raise self._to_error(parsed["error"])
        if parsed.get("ok") is True:
            return parsed.get("result")
        if parsed.get("ok") is False:
            raise self._to_error(parsed.get("error"))
        raise SunlightError("NETWORK_ERROR", cause=ValueError("malformed response"))

    @staticmethod
    def _to_error(err) -> SunlightError:
        code = err.get("code") if isinstance(err, dict) else None
        head = err.get("head") if isinstance(err, dict) else None
        if not isinstance(head, dict) or "seq" not in head or "hash" not in head:
            head = None
        return SunlightError(code if code in ERROR_CODES else "NETWORK_ERROR", head=head)
